### ISSUE TITLE
Hanging when unmounting root file system

### ISSUE THREAD

**Issue by hukoyu** [lsds/sgx-lkl#414]
This test started failing a lot in the couple days. This is a regression.
Log file:
[Standard_Console_Output (7).log](https://github.com/lsds/sgx-lkl/files/4755156/Standard_Console_Output.7.log)

```
[   SGX-LKL  ] SGX-LKL (OE) 0.2.0-dev (404916d) LKL 5.3.0 [DEBUG build (-O0)] [SOFTWARE DEBUG]
[   SGX-LKL  ] nproc=8 ETHREADS=8 CMDLINE="mem=32M"
[   SGX-LKL  ] HW TLS support: conf->fsgsbase=0
[   SGX-LKL  ] Registering disk 0 (path='sgxlkl-python.img', mnt='/', [RW   ])
[   SGX-LKL  ] No tap device specified, networking will not be available.
[   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl-debug/bin/../lib/libsgxlkl.so.signed
[   SGX-LKL  ] oe_create_enclave... result=0 (OE_OK)
[   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
[[  SGX-LKL ]] wg0 has public key ofGuDJNLcGbASaCn2UVMqAIPxwJScjxxLCUPPcI92xo=
make: *** [run-sw] Terminated
```
Pipeline test:
https://tfsproduks1.visualstudio.com/Ab5352d9a-5ed5-48f2-b767-7f273b55b53c/sgx-lkl/_TestManagement/Runs?runId=10178&_a=resultSummary&resultId=100021

**Comment by letmaik** [comment]
@hukoyu The log you pasted in the issue description isn't useful. The attached stdout log is. It would make more sense to instead include that in the text.

**Comment by SeanTAllen** [comment]
Can you recreate this locally? I'm unable to do despite running for 30 minutes. I believe there's a problem but without being able to recreate, it is going to be pretty impossible to find a cause.

**Comment by SeanTAllen** [comment]
I have a hard time reproducing this locally its very rare. However, it should impact on more than just the Python test. I'm still tracking down the final source but,  this calls to lkl_sys_umount in lkl_umount_timeout never returns very rarely.

https://github.com/lsds/lkl/blob/583b1ceb8b67fa469cbb4c222224abf0f1e6529f/tools/lkl/lib/fs.c#L276

My assumption is that when trying to execute the system call that one of the locks that touched by calls to `lkl_syscall` is causing a deadlock but I don't have any evidence of that yet beyond it being the most likely scenario.

**Comment by jxyang** [comment]
I instrumented the kernel to print out message before and after `lkl_sys_umount`, and it shows even though the umount was finished in time, the hang still happens. Now I strongly suspect umount is not the cause of the hang, rather some events after it caused the hang.

Note I couldn't reproduce the failure locally either. I had to run mocked pipeline jobs to repro.
The logs are [here](https://dev.azure.com/sgx-lkl/sgx-lkl/_build/results?buildId=759&view=logs&j=4dfb742b-8e24-5ebe-1c3b-99535daead16&t=ba6e03fa-335e-5c86-d35e-1ceb000f4d73&l=190) and [here](https://dev.azure.com/sgx-lkl/sgx-lkl/_build/results?buildId=760&view=logs&j=9c2dc590-b4fb-502e-6ee8-809bf7e59520&t=715028f2-1245-5d32-8111-ee87353c6a98&l=167)

In the code below, both hangs show we have passed the point of umounting, but get stuck between ` lkl_virtio_netdev_remove` and the end of this function.

```
    // Unmount disks
    long res;
    for (int i = num_disks - 1; i >= 0; --i)
    {
        if (!disks[i].mounted)
            continue;

        SGXLKL_VERBOSE("calling lkl_umount_timeout(%s)\n", disks[i].mnt);
        res = lkl_umount_timeout(disks[i].mnt, 0, UMOUNT_DISK_TIMEOUT);
        if (res < 0)
        {
            sgxlkl_warn(
                "Could not unmount disk %d, %s\n", i, lkl_strerror(res));
        }

        // Root disk, no need to remove mount point ("/").
        if (i == 0)
            continue;

        // Read-only root disk, cannot remove pre-existing mount points for
        // secondary disks.
        if (disks[0].ro)
            continue;

        // Not really necessary for mounts in /mnt since /mnt is
        // mounted as tmpfs itself, but it is also possible to mount
        // secondary images at any place in the root file system,
        // including persistent storage, if the root file system is
        // writeable. For simplicity, remove all mount points here.
        res = lkl_sys_rmdir(disks[i].mnt);
        if (res < 0)
        {
            sgxlkl_warn(
                "Could not remove mount point %s\n",
                disks[i].mnt,
                lkl_strerror(res));
        }
    }

    SGXLKL_VERBOSE("calling lkl_virtio_netdev_remove()\n");
    lkl_virtio_netdev_remove();

    SGXLKL_VERBOSE("calling lkl_sys_halt()\n");
    res = lkl_sys_halt();
    if (res < 0)
    {
        sgxlkl_fail("LKL halt, %s\n", lkl_strerror(res));
    }

    /* Notify host about the guest shutdown */
    sgxlkl_host_shutdown_notification();

    /* Set termination flag to notify lthread scheduler to bail out. */
    lthread_notify_completion();

    lthread_detach2(lthread_self());
    SGXLKL_VERBOSE("lthread_detach2() done\n");

    /* Free the shutdown semaphore late in the shutdown sequence */
    sgxlkl_host_ops.sem_free(termination_sem);

    sgxlkl_free_config(sgxlkl_enclave);

    lthread_exit(NULL);

    SGXLKL_VERBOSE("done\n");
}
```

**Comment by prp** [comment]
@jxyang we noticed that we're running the CI pipeline with 8 ethreads, even though the VMs only have 4 cores. In local testing, you could also try running with oversubscribed ethreads, and see if this makes the issue more likely. Note that you may have to increase the TCS count in the config file before doing this.

**Comment by SeanTAllen** [comment]
I've tried oversubscribing ETHREADS. 16 ethreads, 8 nprocs.

There's no change for me. I still can't recreate locally based on just debug build and run-sw.

**Comment by SeanTAllen** [comment]
I ran this for an hour and was unable to reproduce except for 1 scenario.

I can consistently recreate this bug on the first time I run on a given machine after reboot.
In order to recreate, I need to reboot after a hang.

After the first run (which seems to often hang for me), it will not hang/deadlock.

Note however, that after reboot, I have to 

```bash
make clean
make run-sw
```

without the `make clean`, I don't get the hang. However, given how expensive testing the conditions here is, I've only done about 10 tests with and without `make clean`. 

I wouldn't consider the reboot scenario "reproducible" as it only happened 50% of the time for me and that was mostly in the first half of testing.

Previously, I saw able to consistently create a hang on shutdown for the Python app where it was done in non-release mode, without the kernel module installed where it would only fail if you do make clean before doing make run-hw or make run-sw. I can't reproduce that anymore. I mention it here as it looks somewhat similar to what I am getting occasionally now with the reboot scenario.

If the probability of a hang goes up with first run of a VM, then it could explain issues with CI that we don't get locally.
